src/main/java/com/ngs/stash/externalhooks/hook/ExternalMergeCheckHook.java

```diff
@@ -0,0 +1,222 @@
+        Repository repo = pr.getToRef().getRepository();
+
+        // compat with Stash < 3.2.0
+        String repoPath = this.properties.getRepositoryDir(repo).getAbsolutePath();
+
+        Settings settings = context.getSettings();
+        List<String> exe = new LinkedList<String>();
+        exe.add(this.getExecutable(
+            settings.getString("exe"),
+            settings.getBoolean("safe_path", false)).getPath());
+
+        if (settings.getString("params") != null) {
+            for (String arg : settings.getString("params").split("\r\n")) {
+                exe.add(arg);
+            }
+        }
+
+        StashUser currentUser = authCtx.getCurrentUser();
+        ProcessBuilder pb = new ProcessBuilder(exe);
+
+        Map<String, String> env = pb.environment();
+        env.put("STASH_USER_NAME", currentUser.getName());
+        if (currentUser.getEmailAddress() != null) {
+            env.put("STASH_USER_EMAIL", currentUser.getEmailAddress());
+        } else {
+            log.error("Can't get user email address. getEmailAddress() call returns null");
+        }
+        env.put("STASH_REPO_NAME", repo.getName());
+
+        boolean isAdmin = permissions.hasRepositoryPermission(
+            currentUser, repo, Permission.REPO_ADMIN);
+        env.put("STASH_IS_ADMIN", String.valueOf(isAdmin));
+
+        RepositoryCloneLinksRequest.Builder cloneLinksRequestBuilder =
+            new RepositoryCloneLinksRequest.Builder();
+
+        cloneLinksRequestBuilder.repository(repo);
+
+        RepositoryCloneLinksRequest cloneLinksRequest =
+            cloneLinksRequestBuilder.build();
+
+        Set<NamedLink> cloneLinks = this.repoService.getCloneLinks(
+            cloneLinksRequest
+        );
+
+        for (NamedLink link : cloneLinks) {
+            env.put(
+                "STASH_REPO_CLONE_" + link.getName().toUpperCase(),
+                link.getHref()
+            );
+        }
+
+        env.put(
+            "STASH_BASE_URL",
+            this.properties.getBaseUrl().toString()
+        );
+
+        env.put("STASH_PROJECT_NAME", repo.getProject().getName());
+        env.put("STASH_PROJECT_KEY", repo.getProject().getKey());
+
+        String summaryMsg = "Merge request failed";
+
+        pb.directory(new File(repoPath));
+        pb.redirectErrorStream(true);
+        try {
+            Process process = pb.start();
+            InputStreamReader input = new InputStreamReader(
+                process.getInputStream(), "UTF-8");
+            OutputStream output = process.getOutputStream();
+
+            output.write((
+                          pr.getFromRef().getRepository().getSlug() + " " +
+                          pr.getFromRef().getLatestChangeset() + " " +
+                          pr.getToRef().getRepository().getSlug() + " " +
+                          pr.getToRef().getLatestChangeset() + "\n"
+                          ).getBytes("UTF-8"));
+            output.close();
```

Comment:
> For all of the information available in the pull request, would it be useful to put a stream of XML or JSON on stdin after the "standard" input? It seems kind of clunky to me to try to use environment variables or individual input fields on stdin to describe the information because it is structured.

Is it possible to get s chunk of xml/json generated via some Stash/Bitbucket API function? If there is, then yeah because you don't worry in the future about new data points being added.

If not, but also to keep it consistent with the existing practices, and also thinking that many git hooks are plain shell scripts which would now need access to a json/xml parser, I vote for clunky, and still as env vars :) I also thought about it last night and it's not arbitrary the split between env and stdin vars. It's adherence to the standard git hooks https://git-scm.com/docs/githooks. So you have the chance now to say that you create a new "standard", the "pre-merge" or "merge-request" or "pull-request" hook which takes more than those 3 vars on stdin, **or you stick to the pre-receive standard and put those vars in the env**. You can also just write the JSON/XML bulk to a temp file, export the filename, run the hook, delete the temp file.